### PR TITLE
Add prefill sampling support to TTT models

### PR DESCRIPTION
**Pull request by sraizada-tt**
### Ticket
[Link to Github Issue](https://github.com/tenstorrent/tt-metal/issues/32025)

models CI: https://github.com/tenstorrent/tt-shield/actions/runs/20488654314/job/58877857584

### CODE REVIEW

File: models/common/sampling/generator.py [hunk 1]
```diff
@@ -117,6 +117,7 @@ def reset_sampling_params(self, sampling_params):
         self.tt_penalties.reset_params(
             sampling_params.presence_penalty, sampling_params.frequency_penalty, sampling_params.repetition_penalty
         )
+        #
```
Comment (bgoelTT) [inline models/common/sampling/generator.py]:
```suggestion
```

File: models/tt_transformers/tt/generator.py [hunk 2]
```diff
@@ -331,6 +381,12 @@ def prefill_forward_text(
                 if "image_grid_thw" in local_kwargs:
                     local_kwargs["image_grid_thw"] = local_kwargs["image_grid_thw"][idx]
 
+            # Sampling during prefill is not currently supported with tracing; fall back to no-trace.
+            if sampling_enabled:
+                sampling_executed = True
+                sampling_params_per_out[idx] = _broadcast_formatted_sampling_params(formatted_sampling_params, idx)
+                prompt_tokens_per_out[idx] = prefill_ids[:, :seq_len].repeat(32, 1)
```
Comment (bgoelTT) [inline models/tt_transformers/tt/generator.py]: Is this the workaround you mentioned where prefill traced sampling was interfering with decode traced sampling?

File: models/tt_transformers/tt/model.py [hunk 3]
```diff
@@ -154,7 +159,7 @@ def process_logits_after_prefill_trace(self, logits, last_token_idx):
         if self.model_config["LM_HEAD_INPUT_MEMCFG"].is_sharded():
             logits = ttnn.interleaved_to_sharded(logits, self.model_config["LM_HEAD_INPUT_MEMCFG"])
         logits = self.lm_head(logits)
-        logits = ttnn.to_layout(logits, layout=ttnn.ROW_MAJOR_LAYOUT, memory_config=ttnn.DRAM_MEMORY_CONFIG)
+        # logits = ttnn.to_layout(logits, layout=ttnn.ROW_MAJOR_LAYOUT, memory_config=ttnn.DRAM_MEMORY_CONFIG)
```
Comment (bgoelTT) [inline models/tt_transformers/tt/model.py]: Remove comment

File: models/tt_transformers/tt/generator.py (hunk 2 above)
Comment (sraizada-tt) [reply models/tt_transformers/tt/generator.py]: nope, sorry this comment is not true, I will remove it

File: models/tt_transformers/tt/generator.py [hunk 4]
```diff
@@ -358,23 +416,63 @@ def prefill_forward_text(
...
+                        self.model[model_id],
+                        sampling_params=per_request_params,
+                        prompt_tokens=prompt_tokens_per_out[idx],
+                    )
+                    res["logits"] = ttnn.to_device(
+                        res["logits"], self.model[model_id].mesh_device, memory_config=ttnn.DRAM_MEMORY_CONFIG
+                    )
+                    tt_tokens, tt_log_probs = self.model[model_id].sampling_prefill.sample(
+                        res["logits"],
+                        enable_trace=False,
```
Comment (rdraskicTT) [inline models/tt_transformers/tt/generator.py]: Can we enable trace for sampling like we do it for decode? Maybe `if res["sampling"]:` can be hidden inside `self._easy_trace_prefill`. I assume the problem is to have a fixed allocation of logits tensor since it's being reallocated for slicing in `process_logits_after_prefill_trace`, will https://github.com/tenstorrent/tt-metal/pull/34924 help with this?

Comment (sraizada-tt) [reply models/tt_transformers/tt/generator.py]: tracing is not giving a perf boost, I would only like to enable it when slicing is also traced because I want to avoid a excecute_trace -> no trace -> excecute_trace situation especially if it isn't helping with perf

Comment (rdraskicTT) [reply models/tt_transformers/tt/generator.py]: I agree, should we push slicing PR to happen before this one or should we do a followup for trace?

Comment (rdraskicTT) [reply models/tt_transformers/tt/generator.py]: I'm asking because perf hit for batch=128 is significant, I saw one of the runs and it's more than 50% slower.

Comment (sraizada-tt) [reply models/tt_transformers/tt/generator.py]: that was missing one more optimization, with that I am seeing 3 ms hit per user i.e. 96 ms total for a batch of 128

Comment (sraizada-tt) [reply models/tt_transformers/tt/generator.py]: I think we should follow up later, because I don't know if we can merge both by tomorrow

Comment (rdraskicTT) [reply models/tt_transformers/tt/generator.py]: Ah great, in that case I'm ok with leaving this for a followup

**Review by rdraskicTT — APPROVED** [review]

**Review by bgoelTT — APPROVED** [review]
